subcommand/health-sync/command.go

```diff
@@ -0,0 +1,297 @@
+	syncChecksInterval = 1 * time.Second
+)
+
+type Command struct {
+	UI     cli.Ui
+	config *config.Config
+	log    hclog.Logger
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	sigs   chan os.Signal
+	once   sync.Once
+
+	checks           map[string]*api.HealthCheck
+	dataplaneMonitor *dataplaneMonitor
+	watcherCh        <-chan discovery.State
+
+	// Following fields are only needed for unit tests
+
+	// health-sync signals to this channel whenever it has completed
+	// all the prerequisites before entering the reconciliation loop.
+	doneChan chan struct{}
+
+	// health-sync waits for someone to signal to this channel before
+	// entering the checks reconcilation loop.
+	proceedChan chan struct{}
+
+	// Indicates that the command is run from a unit test
+	isTestEnv bool
+}
+
+func (c *Command) init() {
+	c.ctx, c.cancel = context.WithCancel(context.Background())
+	c.sigs = make(chan os.Signal, 1)
+}
+
+func (c *Command) Run(args []string) int {
+	c.once.Do(c.init)
+	if len(args) > 0 {
+		c.UI.Error(fmt.Sprintf("unexpected argument: %s", args[0]))
+		return 1
+	}
+
+	conf, err := config.FromEnv()
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("invalid config: %s", err))
+		return 1
+	}
+	c.config = conf
+
+	c.log = logging.FromConfig(c.config).Logger()
+	c.dataplaneMonitor = newDataplaneMonitor(c.ctx, c.log)
+
+	if err := c.realRun(); err != nil {
+		c.log.Error("error running main", "err", err)
+		return 1
+	}
+
+	return 0
+}
+
+func (c *Command) realRun() error {
+	signal.Notify(c.sigs, syscall.SIGTERM)
+	defer c.cleanup()
+
+	go c.dataplaneMonitor.run()
+
+	taskMeta, err := awsutil.ECSTaskMetadata()
+	if err != nil {
+		return err
+	}
+
+	clusterARN, err := taskMeta.ClusterARN()
+	if err != nil {
+		return err
+	}
+
+	serverConnMgrCfg, err := c.config.ConsulServerConnMgrConfig(taskMeta)
+	if err != nil {
+		return fmt.Errorf("constructing server connection manager config: %w", err)
+	}
+
+	watcher, err := discovery.NewWatcher(c.ctx, serverConnMgrCfg, c.log)
+	if err != nil {
+		return fmt.Errorf("unable to create consul server watcher: %w", err)
+	}
+
+	go watcher.Run()
+	defer watcher.Stop()
+
+	state, err := watcher.State()
+	if err != nil {
+		return fmt.Errorf("unable to fetch consul server watcher state: %w", err)
+	}
+
+	consulClient, err := c.setupConsulAPIClient(state)
+	if err != nil {
+		return fmt.Errorf("unable to setup Consul API client: %w", err)
+	}
+
+	if !c.isTestEnv {
+		c.watcherCh = watcher.Subscribe()
+	}
+
+	var healthSyncContainers []string
+	healthSyncContainers = append(healthSyncContainers, c.config.HealthSyncContainers...)
+	healthSyncContainers = append(healthSyncContainers, config.ConsulDataplaneContainerName)
+	currentHealthStatuses := make(map[string]string)
+
+	c.checks, err = c.fetchHealthChecks(consulClient, taskMeta)
+	if err != nil {
+		return fmt.Errorf("unable to fetch checks before the reconciliation loop %w", err)
+	}
+
+	if c.isTestEnv {
+		close(c.doneChan)
+		<-c.proceedChan
+	}
+
+	for {
```

Comment:
The entire reconciliation loop was just a copy paste from https://github.com/hashicorp/consul-ecs/pull/207